### PR TITLE
[improve][fn] Support processingGuarantees "EFFECTIVELY_ONCE" in python function 

### LINKED ISSUE #18903: Support processingGuarantees "EFFECTIVELY_ONCE" in python function
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Motivation

Make pulsar python functions support to set `processingGuarantees` to `EFFECTIVELY_ONCE`

### Solution

according to this [docs](https://pulsar.apache.org/docs/next/functions-concepts/#processing-guarantees-and-subscription-types), when the `processingGuarantees` is `EFFECTIVELY_ONCE`, we need to set the subscription type to `failover`.

```python
if self.instance_config.function_details.processingGuarantees == Function_pb2.EFFECTIVELY_ONCE:
    mode = pulsar._pulsar.ConsumerType.Failover
```

### Alternatives

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by tpiperatgod**
Fixes #18903 

Master Issue: #xyz

### Motivation

Make pulsar python functions support to set `processingGuarantees` to `EFFECTIVELY_ONCE`

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [x] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/tpiperatgod/pulsar/pull/9

### CODE REVIEW

**Review by freeznet — APPROVED** [review]

**Review by nlu90 — COMMENTED** [review]
@tpiperatgod Could you explain how these changes together ensure the `effectively_once` processing guarantee?

**Comment by tpiperatgod** [comment]
> @tpiperatgod Could you explain how these changes together ensure the `effectively_once` processing guarantee?

Sorry for not finishing it before. It is currently available for review.

this PR does the following
1. maintain a `message id - producer obj` map during runtime
2. when a message comes in, generate the `producer_name`, `producer_id` and the `sequence_id` based on the message information
3. use `producer_id` to check if the corresponding producer obj exists: if it exists, set `self.producer` to that producer obj; otherwise, create a new producer obj using `producer_name`
4. assign the `sequence_id` when sending the message

**Review by shibd — CHANGES_REQUESTED** [review]
Sorry, the current Java instance `EFFECTIVELY_ONCE` impl maybe have some problems.

We know that `EFFECTIVELY_ONCE` semantics rely on producer message deduplication.

When the function instance produces a message to the sink topic, the source topic message ID will be used as the sequence id.

https://github.com/apache/pulsar/blob/82237d3684fe506bcb6426b3b23f413422e6e4fb/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java#L336-L338

https://github.com/apache/pulsar/blob/17c58a539315cc4ea39655d4328c5caf55f87d3d/pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionCommon.java#L316-L326

**Producer deduplication requires sequenceIds to be ascending**

In case, the source topic has 3 partitions, and the sink topic has 2 partitions and 1 function instance.

```
<----------Order of consumption------------
Source Topic P1: [msgId: 10], [msgId: 11], [msgId: 12]
Source Topic P2: [msgId: 20], [msgId: 21], [msgId: 22]
Source Topic P3: [msgId: 30], [msgId: 31], [msgId: 32]
```

The sequence id of messages we may send to sinkTopic P1 is:
```
-----------Order of producer------------>
Sink Topic P1: [seqId: 10], [seqId: 21], [seqId: 12(will Dup)]
```

In the `EFFECTIVELY_ONCE` scenario, we may need to ensure that the **source topic **and **sink topic** correspond one-to-one?

```
Source topic P1  --> function instance  --> Sink topic P1
Source topic P2  --> function instance  --> Sink topic P2
Source topic P3  --> function instance  --> Sink topic P3
```

If I'm missing something, correct me, thanks.

**Comment by tpiperatgod** [comment]
Thanks, @shibd, based on your comment, I think the mechanism of the effectively-once semantics should be as follows.

The user needs to ensure that one source topic corresponds to one sink topic, at this time the principle of message delivery is: `source topic -> sink topic`
If partition topic is enabled, it is necessary to ensure that the number of partitions of source topic and sink topic is the same, and ensure that the principle of message delivery follows: `source topic Pn -> sink topic Pn; source topic Pm -> sink topic Pm`

Therefore, before using the effectively-once semantics, we should confirm the following points.
1. deduplication is enabled
2. function has only one source topic and one sink topic
3. if partition topic is enabled, we need to make sure the number of partitions is the same

Please feel free to correct me, and what do you think @freeznet @nlu90

**Comment by shibd** [comment]
Hi, 

> Thanks, @shibd, based on your comment, I think the mechanism of the effectively-once semantics should be as follows.
> 
> The user needs to ensure that one source topic corresponds to one sink topic, at this time the principle of message delivery is: `source topic -> sink topic` If partition topic is enabled, it is necessary to ensure that the number of partitions of source topic and sink topic is the same, and ensure that the principle of message delivery follows: `source topic Pn -> sink topic Pn; source topic Pm -> sink topic Pm`
> 
> Therefore, before using the effectively-once semantics, we should confirm the following points.
> 
> 1. deduplication is enabled
> 2. function has only one source topic and one sink topic
> 3. if partition topic is enabled, we need to make sure the number of partitions is the same
> 
> Please feel free to correct me, and what do you think @freeznet @nlu90

Hi, I agree 1, 2.

For 3, I think even if the source topic and sink topic have the same number of partitions, the function framework also does not guarantee that the data of `source topic [Pn]` will be sent to `sink topic [Pn]`

**Comment by tpiperatgod** [comment]
OK, so how about this PR only supports `effectively-once` semantics for non-partitioned topics? or should I put it into the backlog?

and we may need to tell users to enable `deduplication` through documentation to use the `effectively-once` semantics.

**Comment by shibd** [comment]
> OK, so how about this PR only supports `effectively-once` semantics for non-partitioned topics? or should I put it into the backlog?
> 
> and we may need to tell users to enable `deduplication` through documentation to use the `effectively-once` semantics.

We can discuss how to achieve [3] first. @nlu90 @freeznet Can you take a look at it?

**Comment by freeznet** [comment]
@shibd IIRC the python client do not support `PartitionsRoutingMode.CustomPartition` so it is hard to support [3] in python runtime unless the client supports it first. We could create an issue in `pulsar-client-python` and related pulsar function issue in `pulsar` repo as well. Once it is done, we could work on the partitioned topic support.

As for this PR, it is okay for me to make the effectively_once work only with non-partitioned topics, WDYT @nlu90 @shibd @tpiperatgod ?

File: pulsar-functions/instance/src/main/python/python_instance.py [hunk 1]
```diff
@@ -475,6 +509,22 @@ def get_function_status(self):
...
+    """
+    The prerequisites for enabling effective-once semantics are as follows.
+    1. deduplication is enabled
+    2. set ProcessingGuarantees to EFFECTIVELY_ONCE
+    3. the function has only one source topic and one sink topic
+    4. (unsupported yet) if partitioned topic is enabled, ensure that the number of partitions
+      (of both source and sink topics) is the same
+    """
+    if self.instance_config.function_details.processingGuarantees == \
+            Function_pb2.ProcessingGuarantees.Value('EFFECTIVELY_ONCE') and \
```
Comment (freeznet) [inline pulsar-functions/instance/src/main/python/python_instance.py:522]: Could you please add some log output here when `processingGuarantees` is set to `EFFECTIVELY_ONCE`, but other requirements do not meet?

File: pulsar-functions/instance/src/main/python/python_instance.py [hunk 2]
```diff
@@ -302,14 +310,27 @@ def done_producing(self, consumer, orig_message, topic, result, sent_message):
...
-    if output is not None and self.instance_config.function_details.sink.topic != None and \
+    if output is not None and self.instance_config.function_details.sink.topic is not None and \
             len(self.instance_config.function_details.sink.topic) > 0:
       if self.output_serde is None:
         self.setup_output_serde()
+      if self.effectively_once:
+        if self.contextimpl.get_message_partition_index() is None or \
+                self.contextimpl.get_message_partition_index() >= 0:
+          Log.error("Partitioned topic is not available in effectively_once mode.")
+          return
```
Comment (freeznet) [inline pulsar-functions/instance/src/main/python/python_instance.py:322]: Should raise an error here instead of `return` nothing.

**Review by freeznet — COMMENTED** [review]
Left some comments, PTAL.

**Comment by eolivelli** [comment]
Hello @tpiperatgod 
thank you very much for your contribution.
Can you please raise a discussion on the community mailing list dev@pulsar.apache.org ?
This will help the community to be informed of this important work

**Comment by tpiperatgod** [comment]
> Hello @tpiperatgod thank you very much for your contribution. Can you please raise a discussion on the community mailing list [dev@pulsar.apache.org](mailto:dev@pulsar.apache.org) ? This will help the community to be informed of this important work

sure, I will.

**Comment by tpiperatgod** [comment]
> Hello @tpiperatgod thank you very much for your contribution. Can you please raise a discussion on the community mailing list [dev@pulsar.apache.org](mailto:dev@pulsar.apache.org) ? This will help the community to be informed of this important work

please follow up this issue via the link: https://lists.apache.org/thread/qvm955jsmskqx6lhg2mmzft4745lj7th

**Review by shibd — APPROVED** [review]

**Comment by Anonymitaet** [comment]
Hi @tpiperatgod @freeznet @nlu90 this PR is labeled with `doc-required`. I check the relevant docs and have some questions below. Can you please double-check? Thank you!

1.  Doc for `pulsar-admin` reference is done.

    "Python" has been added for `localrun`, `create`, `update` commands, so we do not need to do anything more for this reference doc.
  
  image

2. [Pulsar Feature Matrix (Client + Function)](https://docs.google.com/spreadsheets/d/1YHYTkIXR8-Ql103u-IMI18TXLlGStK8uJjDsOOA0T20/edit#gid=1588411801&range=48:48) needs to be updated.

    Should the Python column of lines 48, 117, 177 be changed as below?

image

**Comment by tpiperatgod** [comment]
This is fine by me, but as discussed in this PR, this semantic is not fully supported (for partitioned topics)

**Comment by Anonymitaet** [comment]
@tpiperatgod thanks for your feedback! 

Double-check: should I make it more accurate as below?

- [pulsar-admin reference](https://pulsar.apache.org/reference/#/next/pulsar-admin/functions?id=localrun)

image

- [Pulsar Feature Matrix (Client + Function)](https://docs.google.com/spreadsheets/d/1YHYTkIXR8-Ql103u-IMI18TXLlGStK8uJjDsOOA0T20/edit#gid=1588411801&range=48:48)

image

**Pull request merged**